File: typescript/course-repo/exercise/03-dedup-via-id-reuse/src/sendDouble.ts
/**
 * Call start() TWICE with the same activityId, back-to-back.
 *
 * Simulates an upstream system sending the same logical event twice.
 * Without an ID conflict policy, the second call raises an error.
 *
 * Fix: pass idConflictPolicy: ActivityIdConflictPolicy.USE_EXISTING to both calls
 * so the second one quietly returns the existing handle.
 */
import { ActivityIdConflictPolicy, Client, Connection } from '@temporalio/client';
import { TASK_QUEUE, WEBHOOK_RECEIVER_URL } from './shared';

// TODO: uncomment the ActivityIdConflictPolicy import above once you're ready to set the policy.

async function start(client: Client, eventId: string, label: string) {
  console.log(`[${label}] start activityId=deliver-${eventId}`);
  try {
    const handle = await client.activity.start("deliverWebhook", {
      args: [{ url: WEBHOOK_RECEIVER_URL, payload: { eventId, type: 'dup_test' }, eventId }],
      id: `deliver-${eventId}`,
      taskQueue: TASK_QUEUE,
      startToCloseTimeout: '30 seconds',
      // TODO: add idConflictPolicy: ActivityIdConflictPolicy.USE_EXISTING
      // so the second call returns the existing handle instead of erroring.
    });
    console.log(`[${label}] handle ok (activityId=${handle.activityId})`);
    return handle;
  } catch (err) {
    console.log(`[${label}] FAILED: ${err}`);
    return null;
  }
}

async function main(eventId: string) {
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const h1 = await start(client, eventId, 'call-1');
  // Second call simulates the upstream retry/replay arriving right after the first.
  const h2 = await start(client, eventId, 'call-2');

  if (h1) { await h1.result(); console.log('[call-1] activity completed'); }
  if (h2) { await h2.result(); console.log('[call-2] activity completed'); }

  await connection.close();
}

// Suppress the unused import warning until the TODO is resolved.
void ActivityIdConflictPolicy;

const eventId = process.argv[2] ?? 'evt_dup';
main(eventId).catch((err) => {
  console.error(err);
  process.exit(1);
});

File: java/course-repo/exercise/03-dedup-via-id-reuse/src/main/java/webhooks/SendDouble.java
package webhooks;

import io.temporal.client.ActivityClient;
import io.temporal.client.ActivityClientOptions;
import io.temporal.client.ActivityHandle;
import io.temporal.client.StartActivityOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

/**
 * Calls start() TWICE with the same Activity ID, back-to-back.
 *
 * Simulates an upstream system sending the same logical event twice.
 * Without an ID conflict policy, the second call raises an error because the
 * default conflict policy is FAIL.
 *
 * Fix: set the conflict policy to USE_EXISTING on the StartActivityOptions so
 * the second call quietly returns a handle to the existing run.
 */
public class SendDouble {

  static ActivityHandle<Integer> start(
      ActivityClient client, String eventId, String label) {
    System.out.println("[" + label + "] start activityId=deliver-" + eventId);

    Map<String, Object> payload = new HashMap<>();
    payload.put("eventId", eventId);
    payload.put("type", "dup_test");
    WebhookDelivery req = new WebhookDelivery(Shared.WEBHOOK_RECEIVER_URL, payload, eventId);

    StartActivityOptions options =
        StartActivityOptions.newBuilder()
            .setId("deliver-" + eventId)
            .setTaskQueue(Shared.TASK_QUEUE)
            .setStartToCloseTimeout(Duration.ofSeconds(30))
            // TODO: set the conflict policy to USE_EXISTING here so the second
            // call returns the existing handle instead of erroring. For example:
            // .setIdConflictPolicy(
            //     io.temporal.api.enums.v1.ActivityIdConflictPolicy
            //         .ACTIVITY_ID_CONFLICT_POLICY_USE_EXISTING)
            .build();

    try {
      ActivityHandle<Integer> handle =
          client.start(WebhookActivities.class, WebhookActivities::deliverWebhook, options, req);
      System.out.println("[" + label + "] handle ok (activityId=deliver-" + eventId + ")");
      return handle;
    } catch (Exception err) {
      // The default FAIL policy makes the second start throw
      // io.temporal.client.ActivityAlreadyStartedException.
      System.out.println("[" + label + "] FAILED: " + err);
      return null;
    }
  }

  public static void main(String[] args) {
    String eventId = args.length > 0 ? args[0] : "evt_dup";

    WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
    ActivityClient client =
        ActivityClient.newInstance(
            service, ActivityClientOptions.newBuilder().setNamespace("default").build());

    ActivityHandle<Integer> h1 = start(client, eventId, "call-1");
    // The second call simulates the upstream retry/replay arriving right after the first.
    ActivityHandle<Integer> h2 = start(client, eventId, "call-2");

    if (h1 != null) {
      h1.getResult();
      System.out.println("[call-1] activity completed");
    }
    if (h2 != null) {
      h2.getResult();
      System.out.println("[call-2] activity completed");
    }

    System.exit(0);
  }
}
